Remove OQS default and update BIKE algorithms

kevinmkane opened the pull request:
Now builds with liboqs 0.7.1

kevinmkane committed:
Remove BIKE L1 and OQS default algorithms

Now builds with liboqs 0.7.1

baentsch reviewed (changes requested):
I'm somewhat confused: When glancing over this PR, it strikes me that DILITHIUM seems to be removed from the README.md but not from the code: Is this intentional? Why? How can this happen given this is generated code? Or did you manually create this PR instead of running `oqs-template/generate.py`? In any case, please bring README.md and code in sync. 
The second question is more mundane/not asking for a change: Why this work on an "old" version of oqs-openssh?

kevinmkane commented:
> I'm somewhat confused: When glancing over this PR, it strikes me that DILITHIUM seems to be removed from the README.md but not from the code: Is this intentional? Why? How can this happen given this is generated code? Or did you manually create this PR instead of running `oqs-template/generate.py`? In any case, please bring README.md and code in sync. The second question is more mundane/not asking for a change: Why this work on an "old" version of oqs-openssh?

The only change I made manually is the change to `oqs-template/generate.yml`. All the other changes are the result of running `oqs-template/generate.py`. I don't know why that line about DILITHIUM got removed from the README, but it was done by the script.

As for why to do this: This is in furtherance of my OQS-enabled version of libssh, which currently interoperates with the 7.9 version of OpenSSH (because the v8 fork didn't exist when I started that work). Douglas requested an update to the latest liboqs. After talking with Christian, the idea is to update both OQS-v7.9 and libssh to use liboqs 0.7.1 and get that out there, since that work is basically done, and then take updating libssh to work with OQS-OpenSSH v8 as a longer-term work item.

kevinmkane commented:
I figured out the issue with `README.md`: `oqs-template/README.md/list_sigs.fragment` skips the first element of the `sigs` array, so that `OQSDEFAULT` doesn't appear in the list. Now that it's gone, I can take that skip out and it generates the README correctly.

kevinmkane committed:
Don't skip the first element of the sigs array when generating README.md

This originally skipped OQSDEFAULT when it was present so it wouldn't be in the list, but now that OQSDEFAULT is gone, everything in the sigs array of generate.yml should be present in the README.

christianpaquin commented:
CI tests timed out; @baentsch, suggestions?

baentsch commented:
> Maybe we should take out the without-openssl options to reduce the time.

I don't see a reason why not to take those out in v7.x. It would be sensible in turn to look into why these tests are not in v8.x.

kevinmkane committed:
Remove without-openssl from CircleCI build job

kevinmkane commented:
Still timing out even with the without-openssl job removed.

christianpaquin commented:
I compared the 7.* and 8.* build scripts and it looks like we're using the same config. Hmmm...

baentsch commented:
> Still timing out even with the without-openssl job removed.

Hmm -- According to https://github.com/open-quantum-safe/openssh/runs/4794278722, it was the "with-openssl" job that timed out; so we looked at/took out the wrong job. Further, it seems logical as that job apparently runs a full matrix test of all oqs-sigs x all oqs-kems whereas "without-openssl" only does ecdsa x all oqs-kems.... 

-> What about doing a compromise (for both, "with-openssl" and "without-openssl"): Run only one SIG of each family (x all KEMs)?

christianpaquin commented:
> -> What about doing a compromise (for both, "with-openssl" and "without-openssl"): Run only one SIG of each family (x all KEMs)?

Yes, that would be the thing to do. I wanted to harmonize with the other projects (in particular, with this's projects 8.x branch tests; curious to know why they don't time out there).

baentsch commented:
> I saw the `with-openssl` job was timing out. I thought your suggestion of taking out the `without-openssl` job might give the `with-openssl` job time to complete.

For all I know the "timeout-counting" is on a per-job basis (and the jobs run in parallel).

> 
> Regardless, I'll restore `without-openssl` since that didn't work, and look into the compromise you suggest.

Thanks!

baentsch commented:
> Yes, that would be the thing to do. I wanted to harmonize with the other projects (in particular, with this's projects 8.x branch tests; curious to know why they don't time out there).

Because we have [the option to run much fewer tests](https://github.com/open-quantum-safe/openssh/blob/0c5eac6a3ba249f73bb953745cf109bdf281eafd/oqs-test/try_connection.py#L191-L200) -- and this option is activated for the CCI run in v8.

christianpaquin commented:
> > Yes, that would be the thing to do. I wanted to harmonize with the other projects (in particular, with this's projects 8.x branch tests; curious to know why they don't time out there).
> 
> Because we have [the option to run much fewer tests](https://github.com/open-quantum-safe/openssh/blob/0c5eac6a3ba249f73bb953745cf109bdf281eafd/oqs-test/try_connection.py#L191-L200) -- and this option is activated for the CCI run in v8.

Ah! I only checked the circleci build config, thinking we would disable long-running algs. Yes, we should definitely do that on 7.x too.

kevinmkane committed:
Revert "Remove without-openssl from CircleCI build job"

This reverts commit e232875b1f587e4e14237be8b4d77b48efe8f6db.

kevinmkane committed:
Implement the "do one" strategy for (sigs x kexs) test matrix due to test run time limits

kevinmkane commented:
I've changed to what the "doone" parameter uses in the v8 branch, since that's what its CI script calls. Tests are now passing, now that there are not nearly as many cases.

christianpaquin commented on oqs-test/test_openssh.py:
```diff
@@ -56,17 +56,11 @@ def test_connection():
+            if ('WITH_OPENSSL' in os.environ and os.environ['WITH_OPENSSL'] != 'true') and ('ecdh' in kex_alg):
+                continue
+            # Due to limits on CircleCI test run time, use the "do one" strategy for the (sigs x kexs) test matrix
+            if (kex_alg == kex_algs[0] or sig_alg == sig_algs[0]):
```
`sig_algs[0]` is the non-PQ `'ssh-ed25519` (vs. the v8 branch, where `sigs[0]` is `ssh-falcon512`). Perhaps we should use `sig_algs[1]` instead to test PQ algs for both kex and sig?

kevinmkane replied:
This seems fine, since along one leg the emphasis is testing the kex algorithms, and on the other leg it's the sig algorithm. If we accept this reduced testing matrix, that the fixed algorithm on the other parameter is non-PQ seems fine.

christianpaquin commented on oqs-template/generate.yml:
```diff
@@ -366,17 +355,8 @@ kems:
-      -
-        name: 'BIKE1_L3_FO'
-        oqs_meth: 'OQS_KEM_alg_bike1_l3_fo'
+        name: 'BIKE_L3'
```
Another point of discussion from our dev meeting: is there a reason why BIKE_L1 was removed? It's enabled in v8, so it should work here as well(?)

kevinmkane replied:
That's a mistake. I misread the equivalent change in the v8 branch and removed one too many. Will fix.

kevinmkane committed:
Restore current version of BIKE-L1 mistakenly removed

kevinmkane committed:
Regenerate templated files after BIKE-L1 restoration

christianpaquin committed:
Update README.md
